# Геокодирование домов

По адресам получаем координаты домов как точек. Геокодирование нужно, чтобы визуализировать вакантность на карте. Геокодируем с помощью Nominatim, запущенного локально в Docker. Инструкция по запуску есть в Readme проекта [nominatim-docker](https://github.com/mediagis/nominatim-docker/tree/master/4.4). PBF-файл Москвы для Nominatim взял [отсюда](https://download.openstreetmap.fr/extracts/russia/central_federal_district/). Дома — из набора открытых данных «[Реестр домов по городу Москве](https://xn--80adsazqn.xn--p1aee.xn--p1ai/opendata/export/101)» ФРТ.

In [1]:
import pandas as pd
import requests
import tqdm

pd.set_option("display.max_columns", None)

In [2]:
houses = pd.read_csv("data/export-reestrmkd-77-20240701.zip", sep=";")

In [3]:
houses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34700 entries, 0 to 34699
Data columns (total 60 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   id                               34700 non-null  int64  
 1   region_id                        34700 non-null  object 
 2   area_id                          1318 non-null   object 
 3   city_id                          2641 non-null   object 
 4   street_id                        33566 non-null  object 
 5   shortname_region                 34700 non-null  object 
 6   formalname_region                34700 non-null  object 
 7   shortname_area                   1318 non-null   object 
 8   formalname_area                  1318 non-null   object 
 9   shortname_city                   2641 non-null   object 
 10  formalname_city                  2641 non-null   object 
 11  shortname_street                 33566 non-null  object 
 12  formalname_street 

In [4]:
houses.sample(2)

id                             region_id area_id city_id  \
13595  7946640  0c5b2444-70a0-4932-980c-b4dc0d3f02b5     NaN     NaN   
27508  8647449  0c5b2444-70a0-4932-980c-b4dc0d3f02b5     NaN     NaN   

                                  street_id shortname_region  \
13595  58d3d1be-8935-4e6c-81cd-5f6eb38b5678                г   
27508  fed18d75-b7f6-46ca-ab5f-3ce32ea358b0                г   

      formalname_region shortname_area formalname_area shortname_city  \
13595            Москва            NaN             NaN            NaN   
27508            Москва            NaN             NaN            NaN   

      formalname_city shortname_street formalname_street house_number  \
13595             NaN               ул      Красных Зорь           49   
27508             NaN               ул   Советской Армии            6   

      building block letter                               address  \
13595      NaN   NaN    NaN    г. Москва, ул. Красных Зорь, д. 49   
27508      NaN   NaN    NaN  г. Москва, ул. Советской Армии, д. 6   

                                  houseguid  management_organization_id  \
13595  8201e785-1af5-4fc0-9763-6bc95080ad15                   7820444.0   
27508  78561dfd-c550-84da-1957-db89534dde5e                         NaN   

       built_year  exploitation_start_year           project_type  \
13595         NaN                   1968.0                 индив.   
27508      2013.0                   2013.0  индивидуальный проект   

                house_type is_alarm   method_of_forming_overhaul_fund  \
13595  Многоквартирный дом      Нет                      Не заполнено   
27508  Многоквартирный дом      Нет  На счете регионального оператора   

       floor_count_max  floor_count_min  entrance_count  elevators_count  \
13595             12.0              NaN             1.0              2.0   
27508             14.0             12.0             2.0              8.0   

      energy_efficiency  quarters_count  living_quarters_count  \
13595      Не заполнено              84                    NaN   
27508      Не заполнено             136                  134.0   

       unliving_quarters_count area_total area_residential  \
13595                      NaN    3660,00          3660,00   
27508                      2.0   36186,00         15109,20   

      area_non_residential area_common_property area_land parking_square  \
13595                 0,00                  NaN    435,30            NaN   
27508               623,91             15136,40      0,00           0,00   

       playground  sportsground other_beautification foundation_type  \
13595         498           498                  NaN    Не заполнено   
27508         499           499                  NaN         Свайный   

           floor_type wall_material basement_area    chute_type  chute_count  \
13595    Не заполнено     Смешанные        373,00  Не заполнено          1.0   
27508  Железобетонные    Монолитные       5600,00   Отсутствует          0.0   

      electrical_type  electrical_entries_count heating_type  \
13595     Центральное                       NaN  Центральное   
27508     Центральное                       3.0  Центральное   

                                      hot_water_type cold_water_type  \
13595  Закрытая с приготовлением горячей воды на ЦТП     Центральное   
27508  Закрытая с приготовлением горячей воды на ЦТП     Центральное   

      sewerage_type sewerage_cesspools_volume     gas_type  \
13595   Центральное                       NaN  Центральное   
27508   Центральное                      0,00  Отсутствует   

                   ventilation_type firefighting_type         drainage_type  
13595                  Не заполнено      Не заполнено          Не заполнено  
27508  Приточно-вытяжная вентиляция    Автоматическая  Внутренние водостоки

In [8]:
houses.loc[houses["building"].notna()].head(2)

id                             region_id area_id city_id  \
27  8098046  0c5b2444-70a0-4932-980c-b4dc0d3f02b5     NaN     NaN   
28  8098048  0c5b2444-70a0-4932-980c-b4dc0d3f02b5     NaN     NaN   

                               street_id shortname_region formalname_region  \
27  0fd90ea9-84a9-4a1f-b67e-44a6f70c1ede                г            Москва   
28  0fd90ea9-84a9-4a1f-b67e-44a6f70c1ede                г            Москва   

   shortname_area formalname_area shortname_city formalname_city  \
27            NaN             NaN            NaN             NaN   
28            NaN             NaN            NaN             NaN   

   shortname_street formalname_street house_number building block letter  \
27               ул         1905 года            9        1   NaN    NaN   
28               ул         1905 года            9        2   NaN    NaN   

                                   address  \
27  г. Москва, ул. 1905 года, д. 9, стр. 1   
28  г. Москва, ул. 1905 года, д. 9, стр. 2   

                               houseguid  management_organization_id  \
27  5f0ede13-7339-4888-91e9-0a662ba0f7c1                   8013388.0   
28  d252563a-c369-4967-a136-2066a2bed31b                   8013388.0   

    built_year  exploitation_start_year project_type           house_type  \
27      1961.0                      0.0      П-18/22  Многоквартирный дом   
28      1961.0                   1961.0        II-18  Многоквартирный дом   

   is_alarm   method_of_forming_overhaul_fund  floor_count_max  \
27      Нет  На счете регионального оператора              9.0   
28      Нет  На счете регионального оператора              9.0   

    floor_count_min  entrance_count  elevators_count energy_efficiency  \
27              9.0             1.0              1.0       Не присвоен   
28              9.0             1.0              1.0       Не присвоен   

    quarters_count  living_quarters_count  unliving_quarters_count area_total  \
27              67                   64.0                      3.0    3343,00   
28              73                   65.0                      8.0    3005,00   

   area_residential area_non_residential area_common_property area_land  \
27          2299,00              1044,00              1044,00      0,00   
28          2317,00               688,00               339,90      0,00   

   parking_square  playground  sportsground other_beautification  \
27           0,00         500           500                    0   
28           0,00         500           500                    0   

   foundation_type      floor_type wall_material basement_area  \
27       Ленточный  Железобетонные     Смешанные          0,00   
28       Ленточный  Железобетонные    Монолитные          0,00   

              chute_type  chute_count electrical_type  \
27  На лестничной клетке          1.0     Центральное   
28  На лестничной клетке          1.0     Центральное   

    electrical_entries_count heating_type  \
27                       1.0  Центральное   
28                       1.0  Центральное   

                                       hot_water_type cold_water_type  \
27      Закрытая с приготовлением горячей воды на ЦТП     Центральное   
28  Открытая с отбором сетевой воды на горячее вод...     Центральное   

   sewerage_type sewerage_cesspools_volume     gas_type  \
27   Центральное                      0,00  Центральное   
28   Центральное                      0,00  Центральное   

                ventilation_type firefighting_type         drainage_type  
27  Приточно-вытяжная вентиляция       Отсутствует  Внутренние водостоки  
28  Приточно-вытяжная вентиляция       Отсутствует  Внутренние водостоки

In [6]:
result = []
for _, row in tqdm.tqdm(houses.iterrows()):
    row = row.fillna(value=0)
    house_id = row["id"]
    street = row["formalname_street"]
    house_number = row["house_number"]
    building = row["building"]
    block = row["block"]
    letter = row["letter"]
    city = "Москва"
    
    if not (street and house_number):
        continue
    
    if building:
        house = f"{house_number}с{building}"
    elif block:
        house = f"{house_number}к{block}"
    elif letter:
        house = f"{house_number}{letter}"
    else:
        house = str(house_number)
    
    street_search = f"{house}, {street}"
    
    resp = requests.get(
        "http://localhost:8080",
        params=dict(
            street=street_search,
            city=city
        )
    )
    
    if resp.ok and len(resp.json()) > 0:
        jdata = resp.json()
        lat = jdata[0].get("lat")
        lon = jdata[0].get("lon")
    else:
        lat, lon = None, None
    
    if lat and lon:
        try:
            lat = float(lat)
            lon = float(lon)
        except ValueError:
            lat, lon = None, None
    
    result.append((house_id, lat, lon))

34700it [25:51, 22.36it/s] 


In [7]:
geo = pd.DataFrame(result, columns=["house_id", "lat", "lon"])
geo.head(2)

house_id        lat        lon
0   8929489  55.727066  37.560168
1   8065593  55.727066  37.560168

In [8]:
geo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33543 entries, 0 to 33542
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   house_id  33543 non-null  int64  
 1   lat       31204 non-null  float64
 2   lon       31204 non-null  float64
dtypes: float64(2), int64(1)
memory usage: 786.3 KB


Статистика геокодирования: 34700 домов в наборе → 33543 с адресом хотя бы в виде «улица + дом» (кое-где нет, например, улицы — их не анализируем) → 31204 с координатами (около 90%).

In [9]:
geo.to_csv("data/houses_coords.csv", index=False)